### ISSUE TITLE
ACP: Implement `Future` for `Option<F>`

### ISSUE THREAD

**Issue by nvzqz** [rust-lang/libs-team#197]
# Proposal

## Problem statement

When working with an `Option<F: Future>` in an `async` context, one may want to defer the handling of `None`. This is currently cumbersome/verbose to do.

## Motivation, use-cases

```rs
let x = match f {
    Some(f) => Some(f.await),
    None => None,
};
```

It is not possible to shorten this to `let x = f.map(|f| f.await)` because the `async` context is not available in a closure.

## Solution sketches

```rs
impl<F: Future> Future for Option<F> {
    type Output = Option<F::Output>;

    fn poll(
        self: Pin<&mut Self>,
        cx: &mut Context<'_>,
    ) -> Poll<Self::Output> {
        match self.as_pin_mut() {
            Some(f) => f.poll(cx).map(Some),
            None => Poll::Ready(None),
        }
    }
}
```

This is implemented in https://github.com/rust-lang/rust/pull/109691.

## Links and related work

Prior art:
- [`OptionFuture`](https://docs.rs/futures/0.3/futures/future/struct.OptionFuture.html)
- [`impl Future for Either<L: Future, R: Future>`](https://docs.rs/either/latest/either/enum.Either.html#impl-Future-for-Either%3CL%2C%20R%3E)

**Comment by Nilstrieb** [comment]
cc @rust-lang/wg-async who might have opinions

**Comment by taiki-e** [comment]
`futures` has a future called `OptionFuture` that does the same thing, but there is some open discussion about how it should behave. https://github.com/rust-lang/futures-rs/issues?q=is%3Aissue+is%3Aopen+optionfuture

**Comment by nvzqz** [comment]
@taiki-e a benefit `Option` would have over `OptionFuture` is that properties of the inner future can be inspected before polling/consuming it. We could change `OptionFuture` to expose the underlying `Option` value via `As{Ref,Mut}` conversions. A downside of switching to `Option` is that `FusedFuture` cannot be implemented on it without bumping the MSRV.

**Comment by taiki-e** [comment]
I did *not* mean "we shouldn't implement `Future` for `Option` because `futures` has something similar" but rather "there are open discussions in the `futures`'s prior art about how it should behave".

**Comment by nvzqz** [comment]
Understood, thanks for elaborating. I agree that behavior questions should be resolved before committing to an implementation.

Here are my opinions:
- https://github.com/rust-lang/futures-rs/issues/2457: Transitioning to `None` may be unexpected behavior since there are some futures that are expected to return `Ready` again. It would also be surprising from a functional perspective if this feature is to be considered as a `map` alternative.
- The other issues do not block this feature, since they are about how the `futures` crate would enhance `Option`'s future-ness.

**Comment by nvzqz** [comment]
Another design question for this feature is "should `Option` instead implement `IntoFuture` for symmetry with its `IntoIterator` implementation?"

I think the answer should be "no" since iterating over `Option` is different than polling it. Iterating requires choices over mutability and ownership, whereas polling has only the single choice of `Pin<&mut Self>`.

**Comment by yoshuawuyts** [comment]
cc/ @rust-lang/wg-async

**Comment by nvzqz** [comment]
@oli-obk pointed out to me that similar functionality in an `async` context would be made possible with:
```rs
let x = try { Some(f?.await) };
```
Despite this, I think this feature is worth pursuing due to enabling:
- More flexibility in a generic context.
- Simplicity in a manual `Future` implementation that deals with `Option` fields.

**Comment by nvzqz** [comment]
i was going to bring up this morning the same thing 😄

@thomcc pointed out to me that the `map` operation would be possible with keyword generics, but it is also blocked on async closures

**Comment by traviscross** [comment]
We discussed this at some length today in the WG-async call.  We came to the consensus that:

**WG-async does not want this impl to exist and recommends this ACP be closed.**

Overall, we felt this was a poor workaround for an async `map` (for reasons discussed in the thread above).  We felt that to whatever degree that some solution other than that may be desired (e.g. for expediency) that it could be better addressed with a more precise tool such as a module-level function or a method on `Option` (but we are not advocating those necessarily).

In the sense that the impl proposed here would "reach through" the outer type into the future within, some of us thought this was analogous to the `Try` impls on `Poll` which some view as having been a mistake.

We discussed alternate semantics, such as having `Option` implement `IntoFuture` such that it would return the [`Ready`](https://doc.rust-lang.org/std/future/struct.Ready.html) future, but none of us were immediately able to propose use cases for that outside of testing code, and we were concerned about the desirability of any of the possible semantics for `pin!(None.into_future()).poll(cx)`.

Work on async closures is proceeding at the moment, and we view that as the main path forward here.

**Comment by ppershing** [comment]
I have a feeling that returning `Poll::Ready(None)` might be detrimental for some use-cases. In our codebase we had use-cases where the sane default behavior was to return `Poll::Pending` unconditionally for `Option::None`.
Consider the following pattern where I have conditionally-enabled sources of events (note: code would not compile as it is).


```rust
let conn_primary: Option<TcpStream> = ... // enabled based on config
let conn_secondary: Option<TcpStream> = ... // enabled based on config

loop {
   select! {
      msg_primary = conn_primary.recv() => {...},
      msg_secondary = conn_secondary.recv() => {...},
   }
}
```
Returning `Poll::Ready(None)` on every iteration would basically starve the event loop if `biased=true` and first option is `None`. Even in non-biased scenario this would be just trashing cycles 50% of time.

As a consequence, having `impl<F: Future> Future<Output=Option<F::Output>> for F` might lead to abovementioned code to compile "easier" by suggesting users to e.g. map over the result which might be easy but wrong thing to do.
